### ISSUE TITLE
how to implement `pkg_resources.require('foo>=42')` with `packaging`?

### ISSUE THREAD

**Issue by dimpase** [pypa/packaging#823]
`packaging` and its components is being advertised as a replacement for  deprecated `pkg_resources` from `setuptools`.
OTOH, how does one use `packaging` to implement something like `pkg_resources.require('foo>=42')` ?

It's not clear at all. Does `packaging` lack such a basic feature as checking the version of a module against a constraint from `requirements.txt` ?

**Comment by abravalheri** [comment]
FYI, this is what [setuptools docs](https://setuptools.pypa.io/en/latest/pkg_resources.html) say about replacements:

> Use of pkg_resources is deprecated in favor of [importlib.resources](https://docs.python.org/3.11/library/importlib.resources.html#module-importlib.resources), [importlib.metadata](https://docs.python.org/3.11/library/importlib.metadata.html#module-importlib.metadata) and their backports ([importlib_resources](https://pypi.org/project/importlib_resources), [importlib_metadata](https://pypi.org/project/importlib_metadata)). Some useful APIs are also provided by [packaging](https://pypi.org/project/packaging) (e.g. requirements and version parsing). Users should refrain from new usage of pkg_resources and should work to port to importlib-based solutions.

Note however that some features[^1] are well know to cause severe performance hits at runtime, and so far no one found a way of implementing these features in a performant way. These features are simply deprecated without replacements.

If you are interested in finding out what are the abstract dependencies of an specific distribution, you can try using [`importlib_metadata.requires`](https://importlib-metadata.readthedocs.io/en/latest/api.html#importlib_metadata.requires). However, this will just give you a ["abstract" list of requirements](https://caremad.io/posts/2013/07/setup-vs-requirement/) as specified by the authors, not the concrete requirements that are actually installed (See notes on https://importlib-metadata.readthedocs.io/en/latest/migration.html#pkg-resources-require).

There are a couple of packages that the community developed overtime to extract dependency information from `requirements.txt`, such as [`pip-tools`](https://pypi.org/project/pip-tools/) or [`pipdeptree`](https://pypi.org/project/pipdeptree/). Maybe that is something useful to have a look.

Finally, `pkg_resources` is opensource, so if you really need the exact functionality that it offers right now but don't mind the performance hit, another possibility is to fork `pkg_resources` or copy the relevant parts to a new project.

[^1]: Like [`WorkingSets`](https://setuptools.pypa.io/en/latest/pkg_resources.html#workingset-objects), and any other function that depends on it.

**Comment by dimpase** [comment]
we are not worried about performance, but we don't want to reimplement a parser for `requirements.txt`.

It really sounds like at serious regression that a one-liner like I gave above needs a lot of work to reimplement all of a sudden, and is no longer available in standard Python modules.

**Comment by abravalheri** [comment]
Hi @dimpase, this was never available in standard Python modules (`pkg_resources` has always been a 3rd-Party package and not part of the standard party).

I think this is part of the natural life cycle of open source software. Some packages (like `setuptools` and `pkg_resources`) have been developed by years in a row, and have changed maintainers a couple of times. The maintainers have the right to decide if they are going to use their volunteer/unpaid time to  maintain some features, or if these feature are not worth the trouble/headache[^1].

But is also part of the natural life cycle of open source that other groups of people may pick up the code base and starting maintaining it as a fork.

If you are not interested in forking/maintaining the functionality yourself, please have a look on the third party packages I mentioned[^2], maybe you are going to find something useful for your use case there.

[^1]: In the case of `pkg_resources` the maintainers have been vocal about the upcoming change for years so that users can be prepared when in the future functionality is dropped.

[^2]: There are probably other 3rd party packages elsewhere that might be useful. It might make sense to do some research.

**Comment by dimpase** [comment]
setuptools, as the replacement for distutils, always looked pretty much standard to me.

**Comment by abravalheri** [comment]
> setuptools, as the replacement for distutils, always looked pretty much standard to me.\nAnd isn't pkg_resources a part of setuptools?

Indeed,`pkg_resources` is distributed as part of `setuptools`. However,`setuptools` is not part of Python's standard library. It is a 3rd party package (and by extension `pkg_resources` is a 3rd party package).

Anyway, please feel free to have a look at the links mentioned above or carry own your own research. You may find something useful for your use case.

**Comment by uranusjr** [comment]
A possible replacement is to use `pip check`.

**Comment by dimpase** [comment]
it doesn't look like `pip check` can analyse a particular requirement for consistency - it just cannot take a list of requirements as input.

**Comment by dimpase** [comment]
>  The maintainers have the right to decide if they are going to use their volunteer/unpaid time to  maintain some features

I don't think we can pretend that maintainers crucial Python packages are unpaid volunteers. They are either well-paid employees of tech companies, charged with doing Python work for their employers, or are directly funded by grants to do their Python work. Why, despite being funded by/for Python, they feel like they can break things at will, is beyond me. </rant>

**Comment by ncoghlan** [comment]
There's no direct replacement for the `pkg_resources.require` API because manipulating the working set of packages at runtime is an inherently fragile approach to import path management. There's nothing *anyone* can do to magically make it more reliable, since the fragility is inherent in the very fact that `sys.path` contains different entries at different times during the program's execution, and hence the result of `import` statements may change based on whether they're executed before or after the path adjustments.

When particular APIs and techniques routinely result in faulty software, they get deprecated and discouraged in favour of more reliable alternatives even when they may have worked correctly in a specific subset of cases. Yes, this sucks for the people that were operating in that happy subset, but it results in a generally more reliable Python ecosystem for all of us in the long run.

In this case, the more reliable replacement for `pkg_resources.require` is to use virtual environments to get the Python interpreter itself to set up the desired working set of packages as it starts up, rather than using `pkg_resources` to manipulate `sys.path` after the interpreter has already started.

**Comment by dimpase** [comment]
Thanks, it's very useful. 

@orlitzky - looks like these macros aren't really fixable :-(

**Comment by orlitzky** [comment]
I think I left a comment on the ticket to the effect of "as a last resort, we could always copy & paste the entire implementation." It's stupid but it would keep things working.

And is easier than trying to tell python programmers that pkgconfig exists :)

**Comment by ncoghlan** [comment]
Given the lack of native Windows tooling support in pkgconfig, it often isn't interesting even to Python developers that do know it exists :)

But yes, since `pkg_resources` is open source, folks with use cases where `pkg_resource.require` does work are free to fork and maintain it for the relevant subset of use cases, even though the `setuptools` developers aren't looking to maintaining the feature for use in arbitrary projects anymore.

**Comment by brettcannon** [comment]
I have locked this issue due to the tone turning rather negative while making inaccurate statements about maintainers.
